Refactor select functionality from `util` package to new `utilsaddr` package

jentfoo opened the pull request:
PR is best reviewed commit by commit.  There are two files which are being moved in this PR:
* `lib/utils/addr.go` - The `NetAddr` struct is referenced broadly across the project.  Having it in the `utils` package makes it difficult to avoid potential cyclic import issues with the `util` package.  This came up trying to adapt the feedback in this PR: https://github.com/gravitational/teleport/pull/30184#discussion_r1287482688

In this change it has moved to `lib/utils/utilsaddr/addr.go`.  The new `utilsaddr` package is brief when referenced and avoids the cyclic import risks.

* `lib/utils/net.go` - I looked for other files which may be a good candidate to move to this package.  This one was a good fit, although some functionality was split to the `addr.go` file.  The bulk of the functionality was renamed into a more precisely named `utilsaddr/context.go`

PR for `e` changes to make temporary alias no longer necessary: https://github.com/gravitational/teleport.e/pull/1963

I have added the backport labels since the `NetAddr` is so broadly used, and I am concerned this change will cause friction when backporting.

codingllama commented on lib/utils/addr/addr.go:
```diff
@@ -355,3 +355,15 @@ func ReplaceUnspecifiedHost(addr *NetAddr, defaultPort int) string {
 }
+
+// ClientIPFromConn extracts host from provided remote address.
+func ClientIPFromConn(conn net.Conn) (string, error) {
```
Move func to its own file? It seems disjointed from the rest.

jentfoo replied:
It seemed related in the sense that it's IP parsing.  I can put it in a separate file, but I felt it was related enough to combine.  Let me know your preference

codingllama commented on lib/utils/addr/addr.go:
In the original file/package:

```go
// TODO(jentfoo): Remove once e/ is updated to use utils/addr.
type NetAddr = addr.NetAddr
type ParseAddr = addr.ParseAddr
// (so on, whatever is needed to not break)
```

codingllama reviewed (commented):
LGTM, minus the potential build breakages.

+1 for backporting, so other backports are not impacted.

+1 for smaller, more focused utility packages.

codingllama reviewed (approved):
Thanks for taking the time to avoid breakages.

jentfoo committed:
Refactor: Move `lib/utils/addr.go` to `utils/utilsaddr/addr.go`

This `NetAddr` struct is referenced broadly across the project.  Having it in the `utils` package makes it difficult to avoid potential cyclic import issues with the `util` package.

jentfoo committed:
Move `ClientIPFromConn` from `utils/net.go` to `utilsaddr/addr.go`

This function fits better with `addr.go` than the Context specific functions in the rest of `net.go`.  This change will faciliate moving those functions next.

jentfoo commented:
I updated the package from `addr` to `utilsaddr`.  `netloc` and `netinfo` were also considered, but after discussing with @codingllama we decided to just go with the more descriptive name.

This was needed because there is a lot of `addr` variables that are complicating the refactor.

greedy52 reviewed (approved):
LGTM.

`lib/utils/utilsaddr` still depends on `api/utils` though. what do you guys think if we move `lib/utils/utilsaddr` and `api/utils/addr.go` to `api/utils/utilsaddr`. I am fine with current change too.

jentfoo commented:
> lib/utils/utilsaddr still depends on api/utils though

I will need a follow up PR to remove `addr_alias`.  Just trying to get the build working (which it's not, I still need to figure out the flakey tests and integration tests, the failure is non-obvious to me)

jentfoo committed:
Move `utils/net.go` to `utilsaddr/context.go`

This functionality matches the rest of the address parsing / structuring code in this package.

jentfoo committed:
Move `AddrsFromStrings` and `ToLowerCaseASCII` from `utils.go` to `utilsaddr/addr.go`

jentfoo committed:
Fixes after rebase

zmb3 commented:
/excludeflake TestLimiter

zmb3 commented:
/excludeflake *

jentfoo commented:
Closing, this PR has significantly rotted again.  It does not seem the effort of continual rebasing is worth it in this case.  However I am happy to re-open it if anyone is interested in paring on the Flakey Test failures.
